### ISSUE TITLE
step_filter() with / without roles

### ISSUE THREAD

**Issue by mlduarte** [tidymodels/recipes#316]
`step_filter()` leads to error in train
> Error in { : task 1 failed - "replacement has 3107 rows, data has 1606"

**Example 1: With formula**
```r
data(okc)
rec_okc <- recipe(Class ~ ., okc) %>%
    step_filter(location == "san francisco")
prepped <- prep(rec_okc, training = okc) %>% juice
sum(prepped$location == "san francisco")
sum(prepped$location != "san francisco")
fit_okc <- train(rec_okc %>% step_rm(location), 
                 data = okc, 
                 method = "glm",
                 family = "binomial",
                 trControl = trainControl(method = "repeatedcv", repeats = 1))
```

**Example 2: No formula**

```r
data(okc)
okc$height[c(1, 34234,22322,2,22)] <- NA
rec_okc <- recipe(okc) %>%
    step_factor2string(all_nominal()) %>%
    step_filter(location == "san francisco") %>%
    update_role(diet, height, new_role = "predictor") %>%
    update_role(Class, new_role = "outcome") %>%
    step_medianimpute(height) %>%
    step_mutate(diet = fct_explicit_na(diet, "missing")) %>%
    step_string2factor(has_role("predictor"), -has_type("date"), -has_type("numeric")) %>%
    step_dummy(has_role("predictor"), -has_type("date"), -has_type("numeric"))

prep(rec_okc,  okc, retain= TRUE) %>% juice
fit_okc <- train(rec_okc %>% step_rm(-all_predictors(), -all_outcomes()), 
                 data = okc, 
                 method = "glm",
                 family = "binomial",
                 trControl = trainControl(method = "repeatedcv", repeats = 1))
```

Was also surprised that you cannot use other variables in recipe, once outcome and predictor roles defined in initial formula:, i.e. following also gives error

```r
data(okc)
rec_okc <- recipe(Class ~ diet, okc) %>%
    step_filter(location == "san francisco")
prepped <- prep(rec_okc, training = okc) %>% juice
```

> Error in location == "san francisco" : 
>   comparison (1) is possible only for atomic and list types 

### Session Info:

```r
sessionInfo()
```

> R version 3.5.3 (2019-03-11)
> Platform: x86_64-w64-mingw32/x64 (64-bit)
> Running under: Windows Server >= 2012 x64 (build 9200)
> 
> Matrix products: default
> 
> locale:
> [1] LC_COLLATE=English_Australia.1252  LC_CTYPE=English_Australia.1252    LC_MONETARY=English_Australia.1252
> [4] LC_NUMERIC=C                       LC_TIME=English_Australia.1252    
> 
> attached base packages:
> [1] stats4    grid      stats     graphics  grDevices utils     datasets  methods   base     
> 
> other attached packages:
>  [1] png_0.1-7          DT_0.5             gridExtra_2.3      ROCR_1.0-7         gplots_3.0.1      
>  [6] pROC_1.13.0        party_1.3-3        strucchange_1.5-1  sandwich_2.5-0     zoo_1.8-4         
> [11] modeltools_0.2-22  mvtnorm_1.0-10     yardstick_0.0.2    rsample_0.0.4      recipes_0.1.5.9000
> [16] parsnip_0.0.1      infer_0.4.0        dials_0.0.2        broom_0.5.1        tidymodels_0.0.2  
> [21] stringr_1.4.0      dplyr_0.8.0.1      purrr_0.3.2        readr_1.3.1        tidyr_0.8.3       
> [26] tibble_2.1.1       tidyverse_1.2.1    magrittr_1.5       caret_6.0-81       lubridate_1.7.4   
> [31] knitr_1.21         forcats_0.3.0      dbplyr_1.3.0       dbplot_0.3.0       Hmisc_4.1-1       
> [36] Formula_1.2-3      survival_2.43-3    lattice_0.20-38    glue_1.3.1         e1071_1.7-0       
> [41] DBI_1.0.0          corrplot_0.84      odbc_1.1.6         scales_1.0.0       kableExtra_0.9.0  
> [46] plotly_4.8.0       ggplot2_3.1.0     
> 
> loaded via a namespace (and not attached):
>   [1] utf8_1.1.4          tidyselect_0.2.5    lme4_1.1-20         htmlwidgets_1.3     munsell_0.5.0      
>   [6] codetools_0.2-16    miniUI_0.1.1.1      withr_2.1.2         colorspace_1.4-0    highr_0.7          
>  [11] rstudioapi_0.9.0    bayesplot_1.6.0     rstan_2.18.2        bit64_0.9-7         generics_0.0.2     
>  [16] TH.data_1.0-10      ipred_0.9-8         xfun_0.4            R6_2.4.0            markdown_0.9       
>  [21] rstanarm_2.18.2     bitops_1.0-6        assertthat_0.2.1    promises_1.0.1      multcomp_1.4-10    
>  [26] nnet_7.3-12         gtable_0.2.0        processx_3.2.1      timeDate_3043.102   rlang_0.3.3        
>  [31] splines_3.5.3       lazyeval_0.2.1      ModelMetrics_1.2.2  acepack_1.4.1       checkmate_1.9.1    
>  [36] inline_0.3.15       yaml_2.2.0          reshape2_1.4.3      modelr_0.1.2        tidytext_0.2.0     
>  [41] threejs_0.3.1       crosstalk_1.0.0     backports_1.1.3     httpuv_1.4.5.1      rsconnect_0.8.13   
>  [46] tokenizers_0.2.1    tools_3.5.3         lava_1.6.5          RColorBrewer_1.1-2  ggridges_0.5.1     
>  [51] Rcpp_1.0.1          plyr_1.8.4          base64enc_0.1-3     ps_1.3.0            prettyunits_1.0.2  
>  [56] rpart_4.1-13        haven_2.0.0         cluster_2.0.7-1     data.table_1.12.0   colourpicker_1.0   
>  [61] SnowballC_0.6.0     matrixStats_0.54.0  tidyposterior_0.0.2 hms_0.4.2           shinyjs_1.0        
>  [66] mime_0.6            evaluate_0.12       xtable_1.8-3        tidypredict_0.3.0   shinystan_2.5.0    
>  [71] readxl_1.2.0        rstantools_1.5.1    compiler_3.5.3      KernSmooth_2.23-15  crayon_1.3.4       
>  [76] minqa_1.2.4         StanHeaders_2.18.1  htmltools_0.3.6     later_0.7.5         libcoin_1.0-4      
>  [81] MASS_7.3-51.1       Matrix_1.2-15       cli_1.1.0           gdata_2.18.0        parallel_3.5.3     
>  [86] gower_0.2.0         igraph_1.2.4        pkgconfig_2.0.2     coin_1.3-0          foreign_0.8-71     
>  [91] xml2_1.2.0          foreach_1.4.4       dygraphs_1.1.1.6    prodlim_2018.04.18  rvest_0.3.2        
>  [96] janeaustenr_0.1.5   callr_3.1.1         digest_0.6.18       rmarkdown_1.11      cellranger_1.1.0   
> [101] htmlTable_1.13.1    shiny_1.2.0         gtools_3.8.1        nloptr_1.2.1        nlme_3.1-137       
> [106] jsonlite_1.6        fansi_0.4.0         viridisLite_0.3.0   pillar_1.3.1        loo_2.0.0          
> [111] httr_1.4.0          pkgbuild_1.0.2      xts_0.11-2          shinythemes_1.1.2   iterators_1.0.10   
> [116] bit_1.1-14          class_7.3-15        stringi_1.4.3       blob_1.1.1          latticeExtra_0.6-28
> [121] caTools_1.17.1.1

**Comment by topepo** [comment]
This is a bug (which I'm not really sure how I'm going to fix) and I'll move the issue to the `caret` repo. 

> Was also surprised that you cannot use other variables in recipe, once outcome and predictor roles defined in initial formula:, i.e. following also gives error

Your formula to recipes says that there are only two "ingredients" so that's all that it keeps.

**Comment by EmilHvitfeldt** [comment]
This is related to what will end up happening in https://github.com/tidymodels/recipes/pull/865

**Comment by EmilHvitfeldt** [comment]
The first couple of issues ere fixed in #865. I found a separate issue in caret itself that I filed here: https://github.com/topepo/caret/issues/1293

Lastly, If you want to filter by an variable but not keeping it as a predictor, you can use the `step_rm()` step to remove it after filtering.

``` r
library(recipes)

data(okc, package = "modeldata")
rec_okc <- recipe(Class ~ diet + location, okc) %>%
  step_filter(location == "san francisco") %>%
  step_rm(location)

rec_okc_prepped <- prep(rec_okc, training = okc)
rec_okc_prepped %>% bake(new_data = okc)
#> # A tibble: 59,855 × 2
#>    diet              Class
#>    <fct>             <fct>
#>  1 strictly anything other
#>  2 mostly other      other
#>  3 anything          other
#>  4 vegetarian        other
#>  5 <NA>              other
#>  6 mostly anything   stem 
#>  7 strictly anything other
#>  8 mostly anything   other
#>  9 strictly anything other
#> 10 mostly anything   other
#> # … with 59,845 more rows
```

<sup>Created on 2022-06-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>
